analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs

```diff
@@ -656,6 +656,23 @@ public void TypeOf()
             type = typeof(object);
             type.ToString(); // Compliant
         }
+
+        public void WithDelegate(object asParameter, object asVariable)
+        {
+            if(asParameter == null)
```

Comment:
[Educational] How is checking for null sets the null state for these variables? Shouldn't S2259 only raise an issue if it's certain that the variable null when it's being dereferenced? Shouldn't it be like this:
```
if(asParameter == null)
{
    SomeAction(asParameter.ToString);       // Noncompliant
}
```

Reply:
```
if(asParameter=null) {
  // Here, we're certain that asParametr has Null
} else {
  // Here, we're certain that asParamet has NotNull. Even when it's empty.
}
// Here, we will arrive twice: Once asParametr=Null, once asParametr=NotNull
```


More user-likely scenario is
```
if(arg !=null) {
  DoSomething(arg.ToString)
} else {
  // And here, we're sure arg==Null. Even if the `else` is not present, such state is possible
}
arg.ToString(); // Noncompliant, this should have been in the "if" branch. We arrive twice here.
```